### PR TITLE
Retain fetch actors during reinterpretation

### LINKED ISSUE #1413: DAMLe reinterpretation does not always retain fetch actors
Canton ships only DAML update expressions around (using `GenNode`s as stable, public representations of such expressions), and uses DAMLe for reinterpretation. The implementation of `reinterpret` respects its behavioral contract (copied below), but the contract is too weak for Canton (and any other implementation that ships DAML. In particular, `reinterpret` loses the fetch actors of top-level `Fetch` nodes. Such nodes don't appear when evaluating submitted commands, but appear in Canton for two reasons:

1. participants receive only a projection of the entire transaction. In a projection, a `Fetch` node can appear as a top-level node.
2. transactions produced by the scenario runner can contain top-level `Fetch` nodes. Canton uses the scenario runner for testing.

This could be worked around in Canton itself, but DAMLe is "morally" the right place for this functionality, and every other ledger implementation that ships only `GenNode`s around (@dajmaki is looking into it) is also affected.

Current contract:
```
evaluate(submit(cmds)) = ResultDone(tx) ==> evaluate(reinterpret(txRoots, cmds.ledgerEffectiveTime)) === ResultDone(tx)
 where:
   evaluate(result) = result.consume(pcs, pkgs, keys)
   txRoots = tx.roots.map(id => tx.nodes.get(id).get).toSeq
   tx === tx' if tx and tx' are equivalent modulo a renaming of node and relative contract IDs
```

### PR DESCRIPTION
**Pull request by oggy-**
Closes #1413 
See ticket for description

### CODE REVIEW

**Comment by oggy-** [comment]
@bitonic I had another go, now changing the engine API. You were right, it wasn't too painful. I sneaked in a change to `submit` to also take submitters as a parameter. This made the code a bit more uniform, but also longer-term I think we should build authorization checks into the interpretation - let me know if that's a bad idea.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 1]
```diff
@@ -69,34 +69,41 @@ final class Engine {
     *      because the caller has not provided a required contract instance or package.
     * </li>
     * </ul>
+    *
+    * This method does NOT perform authorization checks; ResultDone can contain a transaction that's not well-authorized.
     */
-  def submit(cmds: Commands): Result[Transaction.Transaction] = {
+  def submit(submitters: Set[Party], cmds: Commands): Result[Transaction.Transaction] = {
```
Comment (bitonic) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala]: This is the wrong place to put this her. `cmds` already contains the submitter. In fact, what are the semantics of this?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 2]
```diff
@@ -1157,14 +1157,14 @@ final case class Compiler(packages: PackageId PartialFunction Package) {
       // actors are either the singleton set of submitter of an exercise command,
       // or the acting parties of an exercise node
       // of a transaction under reconstruction for validation
-      actors: SExpr): SExpr = {
```
Comment (bitonic) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: explain why this is optional now

Comment (oggy-) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: That's a good catch. It's actually never `Some` any more, so I removed it as a parameter now. In general, exercise actors are computed from the controllers in the new versions of LF, and the compiler does not set them any more. I've added a comment to `compileExercise` to this effect.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 3]
```diff
@@ -64,6 +67,9 @@ object Node {
       copy(coid = f(coid))
 
     override def mapNodeId[Nid2](f: Nothing => Nid2): NodeFetch[Cid] = this
+
+    override def requiredAuthorizers: Set[Party] = actingParties.getOrElse(Set.empty)
```
Comment (bitonic) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala]: This is also dangerous -- is it actually correct?

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala [hunk 4]
```diff
@@ -99,9 +102,11 @@ final case class ScenarioRunner(
     }
   }
 
-  private def mustFail(tx: Transaction, committer: Party) = {
+  private def mustFail(tx: Transaction, committers: Set[Party]) = {
     // Update expression evaluated successfully,
     // however we might still have an authorization failure.
+    val committer = if(committers.size == 1) committers.head else crash("there must be exactly one committer")
```
[the remark this answers is not in the document]
Comment (oggy-) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala]: Changed now

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 3 above)
Comment (oggy-) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala]: It's not correct for versions of LF transactions that have authorized fetches, but no no fetch actors. But I'm unsure what to do; let's discuss when you have time.

**Pull request merged**